src/vsr/journal.zig

```diff
@@ -877,78 +877,56 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
-                //     this entry it turns out not to have the right op.
-                //   (This case (and the accompanying unnecessary read) could be prevented by
-                //   storing the op along with the checksum in `prepare_checksums`.)
-                if (slot) |s| {
-                    journal.faulty.set(s);
-                    journal.dirty.set(s);
+                if (message.header.op != op) {
+                    // Possible causes:
+                    // * The prepare was rewritten since the read began.
+                    // * Misdirected read/write.
+                    // * The combination of:
+                    //   * The primary is responding to a `request_prepare`.
+                    //   * The `request_prepare` did not include a checksum.
+                    //   * The requested op's slot is faulty, but the prepare is valid. Since the
+                    //     prepare is valid, WAL recovery set `prepare_checksums[slot]`. But on
+                    //     reading this entry it turns out not to have the right op.
+                    //   (This case (and the accompanying unnecessary read) could be prevented by
+                    //   storing the op along with the checksum in `prepare_checksums`.)
+                    break :reason "op changed during read";
                 }
 
-                journal.read_prepare_log(op, checksum, "op changed during read");
-                callback(replica, null, null);
-                return;
-            }
-
-            if (message.header.checksum != checksum) {
-                // This can also be caused by a misdirected read/write.
-                if (slot) |s| {
-                    journal.faulty.set(s);
-                    journal.dirty.set(s);
+                if (message.header.checksum != checksum) {
+                    // This can also be caused by a misdirected read/write.
+                    break :reason "checksum changed during read";
                 }
 
-                journal.read_prepare_log(op, checksum, "checksum changed during read");
-                callback(replica, null, null);
-                return;
-            }
+                if (!message.header.valid_checksum_body(message.body_used())) {
+                    break :reason "corrupt body after read";
+                }
+                break :reason null;
+            };
 
-            if (!message.header.valid_checksum_body(message.body_used())) {
+            if (error_reason) |reason| {
                 if (slot) |s| {
                     journal.faulty.set(s);
                     journal.dirty.set(s);
                 }
 
-                journal.read_prepare_log(op, checksum, "corrupt body after read");
+                journal.read_prepare_log(op, checksum, reason);
                 callback(replica, null, null);
-                return;
+            } else {
+                callback(replica, message, destination_replica);
```

Comment:
Might redundantly `assert(message.header.checksum == checksum)` here, as control-flow is not entirely trivial